src/blueprint/cors_params.rs

```diff
@@ -0,0 +1,197 @@
+    }
+
+    pub fn allow_methods_to_header(&self) -> Option<(HeaderName, HeaderValue)> {
+        Some((
+            header::ACCESS_CONTROL_ALLOW_METHODS,
+            self.allow_methods.clone()?,
+        ))
+    }
+
+    pub fn allow_headers_to_header(&self) -> Option<(HeaderName, HeaderValue)> {
+        Some((
+            header::ACCESS_CONTROL_ALLOW_HEADERS,
+            self.allow_headers.clone()?,
+        ))
+    }
+
+    pub fn max_age_to_header(&self) -> Option<(HeaderName, HeaderValue)> {
+        Some((
+            header::ACCESS_CONTROL_MAX_AGE,
+            self.max_age.as_ref()?.clone(),
+        ))
+    }
+
+    pub fn expose_headers_to_header(&self) -> Option<(HeaderName, HeaderValue)> {
+        Some((
+            header::ACCESS_CONTROL_EXPOSE_HEADERS,
+            self.expose_headers.as_ref()?.clone(),
+        ))
+    }
+
+    pub fn vary_to_header(&self) -> Option<(HeaderName, HeaderValue)> {
+        let values = &self.vary;
+        let mut res = values.first()?.as_bytes().to_owned();
+        for val in &values[1..] {
+            res.extend_from_slice(b", ");
+            res.extend_from_slice(val.as_bytes());
+        }
+
+        let header_val = HeaderValue::from_bytes(&res).ok()?;
+        Some((header::VARY, header_val))
+    }
+
+    #[allow(clippy::borrow_interior_mutable_const)]
+    pub fn expose_headers_is_wildcard(&self) -> bool {
+        matches!(&self.expose_headers, Some(v) if v == WILDCARD)
+    }
+}
+
+fn ensure_usable_cors_rules(layer: &CorsParams) -> Result<(), ValidationError<String>> {
+    if layer.allow_credentials {
+        let allowing_all_headers = layer
+            .allow_headers
+            .as_ref()
+            .filter(|val| is_wildcard(val))
+            .is_some();
+
+        if allowing_all_headers {
+            Err(ValidationError::new("Invalid CORS configuration: Cannot combine `Access-Control-Allow-Credentials: true` \
+                with `Access-Control-Allow-Headers: *`".into()))?
+        }
+
+        let allowing_all_methods = layer
+            .allow_methods
+            .as_ref()
+            .filter(|val| is_wildcard(val))
+            .is_some();
+
+        if allowing_all_methods {
+            Err(ValidationError::new("Invalid CORS configuration: Cannot combine `Access-Control-Allow-Credentials: true` \
+                with `Access-Control-Allow-Methods: *`".into()))?
+        }
+
+        let allowing_all_origins = layer.allow_origins.iter().any(is_wildcard);
+
+        if allowing_all_origins {
+            Err(ValidationError::new("Invalid CORS configuration: Cannot combine `Access-Control-Allow-Credentials: true` \
+             with `Access-Control-Allow-Origin: *`".into()))?
+        }
+
+        if layer.expose_headers_is_wildcard() {
+            Err(ValidationError::new("Invalid CORS configuration: Cannot combine `Access-Control-Allow-Credentials: true` \
+             with `Access-Control-Expose-Headers: *`".into()))?
+        }
```

Comment:
Need tests in execution spec for invalid headers.

Reply:
done